File: python/naive.py
from multiprocessing import Process, Value

START_BALANCE = 100_000  # grosze
WITHDRAW_AMT = 70_000    # grosze

def withdraw_naive(balance: Value, amount: int) -> None:
    # check-then-act bez synchronizacji
    if balance.value >= amount:
        balance.value -= amount

def run_once() -> int:
    # lock=False -> celowo brak ochrony
    balance = Value("i", START_BALANCE, lock=False)

    p1 = Process(target=withdraw_naive, args=(balance, WITHDRAW_AMT))
    p2 = Process(target=withdraw_naive, args=(balance, WITHDRAW_AMT))
    p1.start(); p2.start()
    p1.join(); p2.join()

    return balance.value

File: rust/src/bin/naive.rs
use std::thread;

const START_BALANCE: i32 = 100_000; // grosze
const WITHDRAW_AMT: i32 = 70_000;   // grosze

fn main() {
    let mut balance: i32 = START_BALANCE;

    // Dwa wątki próbują zrobić check-then-act na współdzielonym stanie.
    // To jest dokładny odpowiednik idei z python/naive.py, ale Rust to zatrzyma na etapie kompilacji.
    let t1 = thread::spawn(|| {
        if balance >= WITHDRAW_AMT {
            balance -= WITHDRAW_AMT;
        }
    });

    let t2 = thread::spawn(|| {
        if balance >= WITHDRAW_AMT {
            balance -= WITHDRAW_AMT;
        }
    });

    let _ = t1.join();
    let _ = t2.join();

    println!("final balance = {}", balance);
}